app/models/merchant.rb

```diff
@@ -1,5 +1,31 @@
 class Merchant < ApplicationRecord
   has_many :items
   has_many :invoices
+  has_many :customers, through: :invoices
+  has_many :transactions, through: :invoices
   validates :name, presence: true
+
+  def self.highest_revenue(quantity)
+    joins(invoices: [:transactions, :invoice_items])
+      .merge(Transaction.success)
+      .group(:id)
+      .select('merchants.*, SUM(invoice_items.quantity * invoice_items.unit_price) AS total_revenue')
+      .order('total_revenue desc')
+      .limit(quantity)
+  end
+
+  def favorite_customer
+    customers.select('customers.*, count(invoices.merchant_id) AS frequency')
+             .joins(invoices: :transactions)
+             .merge(Transaction.success)
+             .group("customers.id")
+             .order('frequency desc').first
+  end
+
+  def customers_with_pending_transactions
+    customers.select('customers.*')
+             .joins(invoices: :transactions)
+             .merge(Transaction.failed)
+             .group("customers.id")
```

Comment:
@meyerhoferc I know this is stale, but consistent with our conversation this afternoon, I believe one of the issues with failed is that it captures all failed transactions (regardless of whether or not there's a subsequent success). Believe this means that the query above will return invoices that have both a failed and successful transaction. Let's discuss more tomorrow!